homeassistant/helpers/script.py

```diff
@@ -121,6 +133,428 @@ class _SuspendScript(Exception):
+
+    async def _async_run(self) -> None:
+        self._log("Running script")
+        try:
+            for self._step, self._action in enumerate(self._script.sequence):
+                if self._stop.is_set():
+                    break
+                await self._async_step()
+        except _StopScript:
+            pass
+        finally:
+            if not self._stop.is_set():
+                self._changed()
+            self._script._runs.remove(self)  # pylint: disable=protected-access
+            self._stopped.set()
+
+    async def async_stop(self) -> None:
+        """Stop script run."""
+        self._stop.set()
+        await self._stopped.wait()
+
+    async def _async_delay_step(self):
+        """Handle delay."""
+        await asyncio.wait(
+            {self._stop.wait()}, timeout=self._start_delay_step().total_seconds()
+        )
+
+    async def _async_wait_template_step(self):
+        """Handle a wait template."""
+
+        @callback
+        def async_script_wait(entity_id, from_s, to_s):
+            """Handle script after template condition is true."""
+            nonlocal done
+            done.set()
+
+        unsub = self._start_wait_template_step(async_script_wait)
+        if not unsub:
+            return
+
+        try:
+            timeout = self._action[CONF_TIMEOUT].total_seconds()
+        except KeyError:
+            timeout = None
+        done = asyncio.Event()
+        try:
+            await asyncio.wait_for(
+                asyncio.wait(
+                    {self._stop.wait(), done.wait()},
+                    return_when=asyncio.FIRST_COMPLETED,
+                ),
+                timeout,
+            )
+        except asyncio.TimeoutError:
+            if not self._action.get(CONF_CONTINUE, True):
+                self._log(_TIMEOUT_MSG)
+                raise _StopScript
+        finally:
+            unsub()
+
+
+class _BackgroundScriptRun(_ScriptRun):
+    """Manage background Script sequence run."""
+
+    async def async_run(self) -> None:
+        """Run script."""
+        self._hass.async_create_task(self._async_run())
+
+
+class _BlockingScriptRun(_ScriptRun):
+    """Manage blocking Script sequence run."""
+
+    async def async_run(self) -> None:
+        """Run script."""
+        try:
+            await asyncio.shield(self._async_run())
+        except asyncio.CancelledError:
+            await self.async_stop()
+            raise
+
+
+class _LegacyScriptRun(_ScriptRunBase):
+    """Manage legacy Script sequence run."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        script: "Script",
+        variables: Optional[Sequence],
+        context: Optional[Context],
+        shared: Optional["_LegacyScriptRun"],
+    ) -> None:
+        super().__init__(hass, script, variables, context)
+        if shared:
+            self._shared = shared
+        else:
+            # To implement legacy behavior we need to share the following "run state"
```

Comment:
I thought that for shared we would just skip the rest of the delay? Did it actually still invoke after the delay was done?

Reply:
Maybe the variable names or the comment need improvement to make what's happening here clearer.

Basically, the current (i.e., previously existing) implementation has _one_ `Script` object whose `async_run` method can be (and is, in the case of automations) called by separate (trigger) tasks.

The new "legacy" implementation still has just one `Script` object whose `async_run` method can be called multiple times, possibly "simultaneously." But to fit into the new class structure (required to truly support independent, simultaneous runs), there is one `_LegacyScriptRun` object _per call_ to `Script.async_run`, where they all share a common "run state" via the first `_LegacyScriptRun` object's `_current` and `_async_listeners` instance variables (accessed via the `_cur` & `_async_listener` properties.)

The ultimate goal is to make calls to `Script.async_run` -> `_LegacyScriptRun.async_run` behave as closely as possible to how calls to `Script.async_run` (by itself) used to behave.

And, to be clear, it's possible for the "user" (e.g., automation) to call `Script.async_run` (and hence, `_LegacyScriptRun.async_run`) when any previous "run" could be either still in the `async_run` method, or in a suspended state (from a `delay` or `wait_template`.)

I can tell you that it was a bit of a challenge to do this. But I'm feeling pretty good about the implementation now. It certainly passes the tests I created beforehand to verify the "legacy" behavior.